### PR TITLE
Socketio extension improvements

### PR DESCRIPTION
**Pull request by alexkander**
This is an implementation of websocket extension.
Features:
- Application base class for ServerIOServer implementation only
- Documentation to implement a websocket server with an existing express instance (like a pure express or RestApplication).
- Decorators to associate controllers with socket namespaces, to associate controller's methods with events and to inject some elements (like socket, message, server or namespace).
- Booter to autoload controllers in the same way that services, repositories and controllers are autoloaded.
- Customizable sequence class to handler events and ways to customize the response (success o fail) for an event.

All this work is based in the next repositories
- https://github.com/raymondfeng/loopback4-example-websocket/
- https://github.com/strongloop/loopback-next/tree/socketio/extensions/socketio

Pending
- Commit example.
- Readme documentation.

Signed-off-by: Alexander Rondon <arondn2@gmail.com>
## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by alexkander** [comment]
@mayank-SFIN571 I have apply the changes of the PR I did into this one. I will wait for your comment.

Regeards

**Comment by mayank-SFIN571** [comment]
@arondn2 great, i am reviewing these. Meanwhile can you please rebase your branch with mine

**Comment by alexkander** [comment]
@mayank-SFIN571 I have made the rebase

File: packages/cli/test/fixtures/copyright/single-package/src/application.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
-// Copyright IBM Corp. 2020. All Rights Reserved.
+// Copyright ACME Inc. 2020. All Rights Reserved.
```
Comment (mayank-SFIN571) [inline packages/cli/test/fixtures/copyright/single-package/src/application.ts]: Is this change intentional?

**Review by mayank-SFIN571 — CHANGES_REQUESTED** [review]

**Comment by mayank-SFIN571** [comment]
Once it is merged I will add default configuration for the HTTP server again.
What do you think @arondn2

**Comment by alexkander** [comment]
> Once it is merged I will add default configuration for the HTTP server again.

I have tried a lot and have not needed a default configuration, so I think it is not necessary. However, we can add something equivalent like in RestServer:

https://github.com/mayank-SFIN571/loopback-next/blob/cf071cb8546b1922b60443f3e86b1aeb8e336f8a/packages/rest/src/rest.server.ts#L216-L222

and implement `resolveRestServerConfig`

File: packages/cli/test/fixtures/copyright/single-package/src/application.ts (hunk 1 above)
Comment (alexkander) [reply packages/cli/test/fixtures/copyright/single-package/src/application.ts]: @mayank-SFIN571 I am not sure why I pushed this, but I already fixed it.

**Comment by alexkander** [comment]
@mayank-SFIN571 this is my first colaborations, and maybe I'll make some mistakes. I am learning.

Howevert, Do you think I have to squash the commits? I thought it should too easier to review the changes with all the commits.

**Comment by mayank-SFIN571** [comment]
> @mayank-SFIN571 this is my first colaborations, and maybe I'll make some mistakes. I am learning.
> 
> Howevert, Do you think I have to squash the commits? I thought it should too easier to review the changes with all the commits.


@arondn2 no worries on that front, we'll squash those while merging them to the base branch.
I was thinking, lets keep the interface for accepting HTTP server configuration and additional socket configuration separate. This will reduce complexity while applying those configurations to the main server.
I'll update those on your branch and merge them today.
